debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlValueConverters.java

```diff
@@ -324,7 +324,7 @@ protected Charset charsetFor(Column column) {
      * @throws IllegalArgumentException if the value could not be converted but the column does not allow nulls
      */
     protected Object convertJson(Column column, Field fieldDefn, Object data) {
-        if (data == null) {
+        if (data == null && !fieldDefn.schema().isOptional()) {
```

Comment:
Reading this same if/else logic in so many places is a bit worrysome. May I suggest to extract the handling of null into one single method instead:

```
// in JdbcValueConverters
protected Object convertNull(Field fieldDefn, Object fallback) {
        if (fieldDefn.schema().isOptional()) {
            return null;
        }

        Object schemaDefault = fieldDefn.schema().defaultValue();

        return schemaDefault != null ? schemaDefault : fallback;
}
```

And then in the individual conversion methods:

```
protected Object toBigDecimal(Column column, Field fieldDefn, Object data) {
    if (data == null) {
        return convertNull(fieldDefn, BigDecimal.ZERO);
    }

    // convert non-null value
```

Reply:
I was thinking about exactly the same thing too - but the problem is that default value is then not run through the following conversions.

Reply:
> that default value is then not run through the following conversions so they are not equivalent

Wouldn't the default value from the schema not already have the right type? Can you give a specific case where converting it would be needed?

Reply:
In fact, we're talking about the default value stored in the _KC_ schema already, so converting it again seems rather dubious, unless I'm missing something?

Reply:
@gunnarmorling Could yuo please check this code? This was my original solution (now simplified a bit) to follow template method pattern
```
    protected Object convertTemplate(Field fieldDefn, Object data, Object fallback, Supplier<?> convert) {
        if (data == null) {
            if (fieldDefn.schema().isOptional()) {
                return null;
            }
    
            Object schemaDefault = fieldDefn.schema().defaultValue();
            return schemaDefault != null ? schemaDefault : fallback;
        }
        return convert.get();
    }

    protected Object convertBooleanAlt(Column column, Field fieldDefn, Object data) {
        return convertTemplate(fieldDefn, data, false, () -> {
            if (data instanceof Boolean) return data;
            if (data instanceof Short) return ((Short) data).intValue() == 0 ? Boolean.FALSE : Boolean.TRUE;
            if (data instanceof Integer) return ((Integer) data).intValue() == 0 ? Boolean.FALSE : Boolean.TRUE;
            if (data instanceof Long) return ((Long) data).intValue() == 0 ? Boolean.FALSE : Boolean.TRUE;
            return handleUnknownData(column, fieldDefn, data);
        });
    }
```